File: Lecture 2/Python/Distribution Network Power Flow/shirmohammadi.py
import numpy as np
import pickle


def shirmohammadi(system):
    """
    Perform power flow calculations using Shirmohammadi's method.
    """

    # Extract the bus data
    number_of_buses = system['buses'].shape[0]
    p_load = system['buses'][:, 1]
    q_load = system['buses'][:, 2]

    # Extract the branch data
    number_of_branches = system['branches'].shape[0]
    from_bus = system['branches'][:, 1].astype(int)
    to_bus = system['branches'][:, 2].astype(int)
    r = system['branches'][:, 3]
    x = system['branches'][:, 4]

    # Initialize the calculation variables
    S_node = - p_load - 1j * q_load
    Z = r + 1j * x
    V = np.ones(number_of_buses, dtype=complex)
    I_node = np.zeros(number_of_buses, dtype=complex)
    J_branch = np.zeros(number_of_branches, dtype=complex)
    V_old = np.zeros(number_of_buses, dtype=complex)
    tolerance = 1e-6
    iteration = 0

    # Main loop
    while np.any(np.abs(V - V_old) > tolerance):

        # Update the iteration variables
        V_old = np.copy(V)
        iteration += 1

        # Calculate the complex node current injections
        for i in range(number_of_buses):
            I_node[i] = np.conj(S_node[i] / V[i])

        # Backward sweep: calculate the complex branch currents
        for i in range(number_of_branches - 1, -1, -1):
            J_branch[i] = - I_node[to_bus[i]]

            for j in range(number_of_branches):
                if to_bus[i] == from_bus[j]:
                    J_branch[i] += J_branch[j]

        # Forward sweep: calculate the complex node voltages
        for i in range(number_of_branches):
            V[to_bus[i]] = V[from_bus[i]] - Z[i] * J_branch[i]

    return V, iteration


# Sample usage
if __name__ == '__main__':
    # Load the system data
    with open("13 bus system.pkl", "rb") as file:
        system = pickle.load(file)
    # Perform the power flow
    V, iteration = shirmohammadi(system)
    print("Node voltages:", np.abs(V))
    print("Number of iterations:", iteration)

File: Lecture 2/MATLAB/Distribution Network Power Flow/shirmohammadi.m
function [V,Iteration] = shirmohammadi(System)

% Extract the bus data

Number_of_Buses = size(System.Buses,1) ;
P_load = System.Buses.P_load ;
Q_load = System.Buses.Q_load ;

% Extract the branch data

Number_of_Branches = size(System.Branches,1) ;
From_Bus = System.Branches.From_Bus ;
To_Bus = System.Branches.To_Bus ;
R = System.Branches.R ;
X = System.Branches.X ;

% Initialize the calculation variables: complex node power injections,
% branch impedances, complex node voltages, complex node current
% injections, complex branch currents, as well as iteration variables

S = - P_load - sqrt(-1)*Q_load ;
Z = R + sqrt(-1)*X ;
V = ones(Number_of_Buses,1) ;
I = zeros(Number_of_Buses,1) ;
J = zeros(Number_of_Branches,1) ;
V_old = 0*V ;
Tolerance = 1e-6 ;
Iteration = 0 ;

% Main loop

while any(abs(V-V_old)>Tolerance)

    % Update the iteration variables

    V_old = V ;
    Iteration = Iteration + 1 ;

    % Calculate the complex node current injections

    for i = 1 : Number_of_Buses

        I(i) = conj(S(i)/V(i)) ;

    end

    % Backward sweep: calculate the complex branch currents

    for i = Number_of_Branches : -1 : 1

        J(i) = - I(To_Bus(i)) ;

        for j = 1 : Number_of_Branches

            if To_Bus(i) == From_Bus(j)

                J(i) = J(i) + J(j) ;

            end

        end

    end

    % Forward sweep: calculate the complex node voltages

    for i = 1 : Number_of_Branches

        V(To_Bus(i)) = V(From_Bus(i)) - Z(i)*J(i) ;

    end

end

end